src/neuroconv/tools/path_expansion.py

```diff
@@ -0,0 +1,77 @@
+import abc
+import os
+from pathlib import Path
+from typing import Iterable, List, Union
+
+from parse import parse
+from pydantic import DirectoryPath, FilePath
+
+from ..utils import DeepDict
+
+
+class AbstractPathExpander(abc.ABC):
+    def extract_metadata(self, folder, format_: str):
+        format_ = format_.replace("\\", os.sep)  # Actual character is a single back-slash; first is an escape for that
+        format_ = format_.replace("/", os.sep)  # our f-string uses '/' to communicate os-independent separators
+        for filepath in self.list_directory(folder):
+            result = parse(format_, filepath)
+            if result:
+                yield filepath, result.named
+
+    @abc.abstractmethod
+    def list_directory(self, folder: Union[FilePath, DirectoryPath]) -> Iterable[str]:
```

Comment:
Also, should this be only a `DirectoryPath`? Iteration fails silently (no return values) if you specify a `FilePath`, perhaps an error would be clearer?

Reply:
Annotation updated - error can be added later if usage becomes a problem